verticapy/learn/causal_model.py

```diff
@@ -0,0 +1,343 @@
+		if show_progress:
+			progress_bar = tqdm(total=max_cond_vars)
+			progress_bar.set_description("Using n conditional variables: 0")
+
+		# Discover the skeleton
+		for cond_set_size in range(max_cond_vars):
+			if method == 'stable':
+				neighbors = {x: skeleton_edges[x].copy() for x in columns}
+				deleted_edges = set()
+				for (x,y) in all_pairs:                
+					for z in itertools.chain(
+											 itertools.combinations(neighbors[x]-set([y]), cond_set_size),
+											 itertools.combinations(neighbors[y]-set([x]), cond_set_size)): # set of all conditionals
+						z = list(z)
+						if cit(self.vdf, x, y, z, alpha = significance_level):
+							deleted_edges.add((x,y))
+							skeleton_edges[x].remove(y)
+							skeleton_edges[y].remove(x)
+							sepsets[frozenset((x,y))] = z
+							break
+				all_pairs = all_pairs - deleted_edges
+		
+			elif method == 'orig':
+				deleted_edges = set()
+				for (x,y) in all_pairs:                
+					for z in itertools.chain(
+											 itertools.combinations(skeleton_edges[x]-set([y]), cond_set_size),
+											 itertools.combinations(skeleton_edges[y]-set([x]), cond_set_size)): # set of all conditionals
+						z = list(z)
+						if cit(self.vdf, x, y, z, alpha = significance_level):
+							deleted_edges.add((x,y))
+							skeleton_edges[x].remove(y)
+							skeleton_edges[y].remove(x)
+							sepsets[frozenset((x,y))] = z
+							break
+				all_pairs = all_pairs - deleted_edges
+			else:
+				raise AttributeError("Choose from valid method")
+			
+			if show_progress:
+				progress_bar.update(1)
+				progress_bar.set_description(f"Using n conditional variables: {cond_set_size + 1}")
+		
+		if show_progress:
+			progress_bar.close()
+
+		self.skeleton = skeleton_edges
+		self.sepsets = sepsets
+
+		return (skeleton_edges, sepsets)
+
+	# ---#
+	def build_pdag(self):
+		"""
+		Builds the partially directed acyclic graph
+
+		Returns
+		-------
+		pdag: dict
+			directed egdes in an adjacency dictionary
+		"""
+		skeleton_edges = self.skeleton
+		sepsets = self.sepsets   
+		
+		# Create partially directed graph
+		columns = skeleton_edges.keys()	
+		pdag = {x: skeleton_edges[x].copy() for x in columns}
+		
+		# 1 Orientate the triples X - Z - Y into X -> Z <- Y  
+		node_pairs = set(itertools.permutations(columns, 2))
+
+		for (x,y) in node_pairs:
+			if y not in skeleton_edges[x]:
+				for z in set(skeleton_edges[x] & skeleton_edges[y]):
+					#if z in pdag[x] and z in pdag[y] and x in pdag[z] and y in pdag[z]:
+					if z not in sepsets[frozenset((x,y))]:
+						if x in pdag[z]:
+							pdag[z].remove(x)
+						if y in pdag[z]:
+							pdag[z].remove(y)
+		
+		edges_to_be_oriented = True
+		while edges_to_be_oriented:
+			remaining_edges = sum([len(pdag[x]) for x in columns])
+		# 2 Orientate non-colliders X -> Z - Y into X -> X -> Y
+			for (x,z) in node_pairs:
+				if z in pdag[x] and x not in pdag[z]:
+					for y in pdag[z]:
+						if z in pdag[y] and y not in skeleton_edges[x]:
```

Comment:
I don't understand the reason of the `y not in skeleton_edges[x]` condition!

Reply:
The intuition behind this rule can be found [here](http://mcb111.org/w06/KollerFriedman.pdf#page=125) but it is basically a stronger statement: if there is no edge between X and Y then we are sure we can direct Z -> Y.